### ISSUE TITLE
autosplitting for entering nether, exiting nether, and entering end

### ISSUE THREAD

**Issue by thorgot** [Jorkoh/LiveSplit.Minecraft#2]
I noticed you added autosplitting for starting and ending the run, which is great! I was working on an ASL script autosplitter a few weeks ago but lost interest before uploading it. I lost interest because it requires unzipping a file, which required calling an outside program like 7zip since autosplit scripts aren't allowed to unzip files. But if you want to unzip within C# I'm pretty sure you can `use ZipFile` instead!

My script:

1. starts if a new save file is detected, like you are doing.
2. splits when the nether region file is created when first entering the nether
3. splits when leaving the nether by using the fact that the player data file updated and the dimension is overworld
4. splits when entering the end when the end region file is created
5. splits when credits are seen, like you are doing
6. resets when the lock file on the current save can be read (indicating the save is no longer loaded, i.e. the player is back on the main menu)

```
state("javaw")
{
}

//NOTE: Assumes 7z.exe can be located in "c:\Program Files\7-Zip\" or "c:\Program Files (x86)\7-Zip\"

startup
{
	refreshRate = 10;
	
	vars.previousCreationTime = new DateTime();
	vars.savePath = "";
	vars.playerDataFile = "";
	vars.lastPlayerDataFileUpdate = new DateTime();
	vars.splitNumber = 0;
	vars.seenCredits = false;
	vars.dirtyDataFile = false;
	vars.dimension = "";
	vars.locationOf7Zip = "";
	vars.saveCreationTimes = new Dictionary<string, DateTime>();
	if (File.Exists("c:\\Program Files\\7-Zip\\7z.exe"))
	{
		vars.locationOf7Zip = "\"c:\\Program Files\\7-Zip\\7z.exe\"";
		//print("Found 7zip at " + vars.locationOf7Zip);
	}
	else if (File.Exists("c:\\Program Files\\7-Zip (x86)\\7z.exe"))
	{
		vars.locationOf7Zip = "\"c:\\Program Files\\7-Zip (x86)\\7z.exe\"";
		//print("Found 7zip at " + vars.locationOf7Zip);
	}
	else
	{
		//print("Unable to find 7zip, nether and victory splits will not trigger!");
	}
	var path = System.IO.Path.Combine(Environment.GetEnvironmentVariable("appdata"), ".minecraft\\saves");
	var saves = System.IO.Directory.EnumerateDirectories (path);
	if (saves.Any())
	{
		foreach (var save in saves)
		{
			if (!vars.saveCreationTimes.ContainsKey(save))
			{
				var advancements = save + "\\advancements";
				if (System.IO.Directory.Exists(advancements))
				{
					vars.saveCreationTimes[save] = System.IO.Directory.GetCreationTime(advancements);
				}
			}

			// advancements folder not yet created so this save is either borked or hasn't been loaded yet
			if (!vars.saveCreationTimes.ContainsKey(save)) continue;

			var creationTime = vars.saveCreationTimes[save];
			if (creationTime > vars.previousCreationTime)
			{
				vars.previousCreationTime = creationTime;
				vars.savePath = save;
				var playerDataFiles = System.IO.Directory.EnumerateFiles(save + "\\playerdata", "*.dat");
				if (playerDataFiles.Any())
				{
					vars.playerDataFile = playerDataFiles.First();
				}
			}
			
		}
	}
	//print("Newest save created at " + vars.previousCreationTime.ToString());

}


start
{

	var newSave = false;
	var path = System.IO.Path.Combine(Environment.GetEnvironmentVariable("appdata"), ".minecraft\\saves");
	var saves = System.IO.Directory.EnumerateDirectories (path);
	if (saves.Any())
	{
		// Start when the advancements folder is created as that is the last thing done while loading a new world
		foreach (var save in saves)
		{
			if (!vars.saveCreationTimes.ContainsKey(save))
			{
				var advancements = save + "\\advancements";
				if (System.IO.Directory.Exists(advancements))
				{
					vars.saveCreationTimes[save] = System.IO.Directory.GetCreationTime(advancements);
				}
				
			}

			// advancements folder not yet created so this save is either borked or hasn't been loaded yet
			if (!vars.saveCreationTimes.ContainsKey(save)) continue;

			var creationTime = vars.saveCreationTimes[save];
			if (creationTime > vars.previousCreationTime) {
				newSave = true;
				vars.previousCreationTime = creationTime;
				vars.savePath = save;
				var playerDataFiles = System.IO.Directory.EnumerateFiles(save + "\\playerdata", "*.dat");
				if (playerDataFiles.Any())
				{
					vars.playerDataFile = playerDataFiles.First();
				}
			}
		}
	}
	
	if (newSave)
	{
		//print("New player data file found at " + vars.playerDataFile);
		
		vars.splitNumber = 0;
		vars.seenCredits = false;
		vars.dirtyDataFile = false;
		vars.dimension = "";
		return true;
	}
	return false;
}

//Extract player .dat file every time it is modified and read seenCredits and Dimension from it
update
{
	if (timer.CurrentPhase != TimerPhase.Running)
	{
		return;
	}

	if (vars.playerDataFile.Length == 0)
	{
		return;
	}
	
	if (vars.locationOf7Zip.Length == 0)
	{
		return;
	}

	var currentPlayerDataFileUpdate = System.IO.Directory.GetLastWriteTime(vars.playerDataFile);
	if (currentPlayerDataFileUpdate == vars.lastPlayerDataFileUpdate)
	{
		return;
	}

	//print("playerData has been updated at " + currentPlayerDataFileUpdate.ToString() + ", last update was at " + vars.lastPlayerDataFileUpdate.ToString());
	vars.lastPlayerDataFileUpdate = currentPlayerDataFileUpdate;
	vars.dirtyDataFile = true;
	
	// Extract current player data file so it can be read
	var extractCommand = "/c \"" + vars.locationOf7Zip + " X \"" + vars.playerDataFile + "\" -o\"" + vars.savePath + "\\playerData\" -y\"";
	//print ("Executing command: " + extractCommand);
	System.Diagnostics.ProcessStartInfo procStartInfo = new System.Diagnostics.ProcessStartInfo("cmd", extractCommand);
	// The following commands are needed to redirect the standard output.
	// This means that it will be redirected to the Process.StandardOutput StreamReader.
	procStartInfo.RedirectStandardOutput = true;
	procStartInfo.RedirectStandardError = true;
	procStartInfo.UseShellExecute = false;
	// Do not create the terminal window.
	procStartInfo.CreateNoWindow = true;
	// Now we create a process, assign its ProcessStartInfo and start it
	System.Diagnostics.Process proc = new System.Diagnostics.Process();
	proc.StartInfo = procStartInfo;
	proc.Start();
	// Get the output into a string for debugging
	//string result = proc.StandardOutput.ReadToEnd();
	string resultError = proc.StandardError.ReadToEnd();
	proc.WaitForExit();
	// Display the command output.
	//print(result);
	if (resultError.Length > 0) print(resultError);
	
	var unzippedDataFile = vars.playerDataFile.Substring(0, vars.playerDataFile.Length - 4);
	//print ("Reading unzipped data file: " + unzippedDataFile);
	if (File.Exists(unzippedDataFile))
	{
		var playerData = File.ReadAllBytes(unzippedDataFile);
		byte[] seenCreditsBytes = Encoding.UTF8.GetBytes("seenCredits");
		byte[] dimensionBytes = Encoding.UTF8.GetBytes("Dimension");
		var seenCreditsLocation = -1;
		for (int i = 0; i < playerData.Length - seenCreditsBytes.Length; i++)
		{
			var segment = new ArraySegment<byte>( playerData, i, seenCreditsBytes.Length ).ToArray();

			//print("Comparing " + Encoding.UTF8.GetString(seenCreditsBytes, 0, seenCreditsBytes.Length) + " to " + Encoding.UTF8.GetString(segment, 0, segment.Length));
			if (segment.SequenceEqual(seenCreditsBytes))
			{
				seenCreditsLocation = i;
				break;
			}
		}
		var dimensionLocation = -1;
		for (int i = 0; i < playerData.Length - dimensionBytes.Length; i++)
		{
			var segment = new ArraySegment<byte>( playerData, i, dimensionBytes.Length ).ToArray();

			//print("Comparing " + Encoding.UTF8.GetString(dimensionBytes, 0, dimensionBytes.Length) + " to " + Encoding.UTF8.GetString(segment, 0, segment.Length));
			if (segment.SequenceEqual(dimensionBytes))
			{
				dimensionLocation = i;
				break;
			}
		}
		//print("File size: " + playerData.Length.ToString());
		//print("Location of seenCredits: " + seenCreditsLocation.ToString());
		//print("Location of Dimension: " + dimensionLocation.ToString());
		if (seenCreditsLocation >= 0)
		{
			//print("Value of seenCredits: " + (playerData[seenCreditsLocation+seenCreditsBytes.Length] == 1));
			vars.seenCredits = playerData[seenCreditsLocation+seenCreditsBytes.Length] == 1;
		}
		if (dimensionLocation >= 0)
		{
			//print("Value of Dimension: " + Encoding.UTF8.GetString(playerData, dimensionLocation+dimensionBytes.Length+2, 19));
			vars.dimension = Encoding.UTF8.GetString(playerData, dimensionLocation+dimensionBytes.Length+2, 19);
		}
	}
	else
	{
		//print ("unzipped data file doesn't exist!");
	}
}

split
{
	int splitNumber = vars.splitNumber;
	switch (splitNumber)
	{
		case 0:
			// First split (Enter Nether) occurs when the nether region file is created
			if (System.IO.Directory.Exists(vars.savePath + "\\DIM-1\\region"))
			{
				vars.splitNumber++;
				vars.dirtyDataFile = false;
				return true;
			}
			break;
		case 1:
			// Second split (Leave Nether) occurs when the player data file is updated (which happens when leaving the nether but not entering it) and has a dimension of overworld
			if (vars.dirtyDataFile && vars.dimension == "minecraft:overworld")
			{
				// Double check that playerdata has been fully extracted to avoid prematurely splitting
				var unzippedDataFile = vars.playerDataFile.Substring(0, vars.playerDataFile.Length - 4);
				var unzippedDataFileUpdateTime = System.IO.Directory.GetLastWriteTime(unzippedDataFile);
				if (unzippedDataFileUpdateTime < vars.lastPlayerDataFileUpdate)
				{
					return false;
				}
				vars.splitNumber++;
				vars.dirtyDataFile = false;
				return true;
			}
			break;
		case 2:
			// Third split (Enter End) occurs when the end region file is created
			if (System.IO.Directory.Exists(vars.savePath + "\\DIM1\\region"))
			{
				vars.splitNumber++;
				vars.dirtyDataFile = false;
				return true;
			}
			break;
		case 3:
			// Last split (Victory) occurs when the player data file is updated to have a seenCredits of 1
			if (vars.seenCredits)
			{
				vars.splitNumber++;
				vars.dirtyDataFile = false;
				return true;
			}
			break;
			
	}
	return false;
}

reset
{
	// Try reading the lock file. If unable, the player is currently running so do not reset.
	var lockfile = vars.savePath + "\\session.lock";
	if (File.Exists(lockfile))
	{
		try
		{
			StreamReader sr = new StreamReader(lockfile);
			sr.Close();
			return true;
		}
		catch (IOException)
		{
			return false;
		}
		catch (Exception)
		{
			throw;
		}
	}
	return false;
}
```

Is any of that logic useful for porting to your autosplitter?

**Comment by Jorkoh** [comment]
Hey, that's pretty cool!

The zipping issues when dealing with NBT files can be solved by just using a NBT library which is significantly faster than calling a external program to unzip it and less error prone than `ZipFile`. I know that you had no choice in ASL and the workaround is pretty good.

Using the lock file to reset wouldn't work since glitched categories rely on restarting the game and even on glitchless categories you are allowed to reload the world to fix bugs like invisible blocks. The splits for Nether and The End when the region files are created sound like a good and reliable idea. I looked into it but decided to keep it simple for various reasons:

1. **Autosplitting with game files is not precise**. You have to keep checking the files which is slow and the game takes time to write stuff on the files. Specially on slow setups with hdds that are recording at the same time people have reported the autosplitter taking several seconds to start the timer after world load simply because it takes that long to update the save files. Those people then want to use that RTA on the leaderboards and it's just a pain.

2. **Autosplitting with game files is error prone**. Different versions do stuff diferently, `seenCredits` is only available on +1.12, IGT is `play_one_minute` now but it was `playOneMinute` before and not part of the stats file before that. You end up having to check different versions and complicate the logic to account for all of them. Even a implementation as simple as the current one is apparently problematic for some users (altough I haven't been able to get a proper bug report yet).

3. **Reading game files is kinda against the rules**. In theory you are not supposed to touch the NBT files at all, the current implementation already pushes the rule and adding more stuff would probably require approval of the mods.

4. **A lot of users are not great with this stuff**. Minecraft speedrunning is pretty popular right now and the audience is often young. Even keeping it as simple as it is now and with a complete video tutorial I get a ton of questions on how to locate the saves folder or why doesn't it work when the game is fullscreened. If I added extra splits people would just check the boxes, don't add splits and complain about the timer stopping when they enter the Nether.

Initially I wanted to add autosplitting (and more stuff) avoiding this issues by using a Fabric mod that would communicate directly with the component, you can check it out in the other branch, but the mod team didn't approve it. I finally had to add the current implementation since it kept being requested. Now an actual speedrunning client is in development so I will hopefully continue the Fabric approach there.

Sorry that this turned into a rant, I really appreciate you sharing your progress but since I most likely won't implement it soon I wanted to give a complete answer to justify it.

**Comment by thorgot** [comment]
All good points. Thanks for reading my suggestion and for the detailed response.
